### PR TITLE
Add new API appservice specific public room list

### PR DESCRIPTION
**Pull request by erikjohnston**
This adds the following:

- A `instance_id` field to each instance in the `GET /thirdparty/protocols` API
- An API that allows ASes to publish rooms to their own dedicated lists, `PUT /directory/list/appservice/$network_id/$room_id`
- Two extra params to ` POST /publicRooms`, `third_party_instance_id` which returns rooms published to AS instance specific lists, and `include_all_networks` which if true returns all rooms that have been published to any list. By default it will only return rooms on the main, default list.

https://docs.google.com/document/d/12mVuOT7Qoa49L_PQAPjavoK9c2nalYEFOHxJOmH5j-w/edit?usp=sharing

### CODE REVIEW

File: synapse/handlers/directory.py [hunk 1]
```diff
@@ -339,3 +339,15 @@ def edit_published_room_list(self, requester, room_id, visibility):
         yield self.auth.check_can_change_room_list(room_id, requester.user)
 
         yield self.store.set_room_is_public(room_id, visibility == "public")
+
+    @defer.inlineCallbacks
+    def edit_published_appservice_room_list(self, appservice_id, network_id,
+                                            room_id, visibility):
+        """Edit the appservice/network specific public room list.
+        """
```
Comment (NegativeMjark) [inline synapse/handlers/directory.py]: Could we have some docstring for the parameters here?

File: synapse/storage/room.py [hunk 2]
```diff
@@ -133,6 +139,73 @@ def set_room_is_public_txn(txn, next_id):
                 set_room_is_public_txn, next_id,
             )
 
+    @defer.inlineCallbacks
+    def set_room_is_public_appservice(self, room_id, appservice_id, network_id,
+                                      is_public):
+        """Edit the appservice/network specific public room list.
```
Comment (NegativeMjark) [inline synapse/storage/room.py]: Could we have docstring for what the params are and what this does?

File: synapse/storage/schema/delta/39/appservice_room_list.sql [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ * limitations under the License.
+ */
+
+CREATE TABLE appservice_room_list(
+    appservice_id TEXT NOT NULL,
+    network_id TEXT NOT NULL,
+    room_id TEXT NOT NULL
+);
+
+CREATE UNIQUE INDEX appservice_room_list_idx ON appservice_room_list(
```
Comment (NegativeMjark) [inline synapse/storage/schema/delta/39/appservice_room_list.sql]: Maybe add comment on what this table is for?

File: synapse/handlers/room_list.py [hunk 4]
```diff
@@ -34,17 +35,38 @@
...
+            limit (int)
+            since_token (str)
+            search_filter (dict)
+            network_tuple (ThirdPartyInstanceID): Which public list to use.
+                This can be (None, None) to indicate the main list, or a particular
+                appservice and network id to use an appservice specific one.
+                Setting to None returns all public rooms across all lists.
+        """
+        if search_filter or network_tuple is not (None, None):
             # We explicitly don't bother caching searches.
```
Comment (NegativeMjark) [inline synapse/handlers/room_list.py]: Does the comment need updating if we aren't caching the appservice lists as well?

**Comment by NegativeMjark** [comment]
What is an appservice_id look like ooi?

**Comment by erikjohnston** [comment]
> What is an appservice_id look like ooi?

e.g. `irc-freenode` or `9666621b268e14d7b18b110073f07c6e955c60a8c8a5792fd810e9978d1931d7`

**Comment by erikjohnston** [comment]
Now with added sytests https://github.com/matrix-org/sytest/pull/328!

**Comment by lukebarnard1** [comment]
Is it really `instance_id`? Because I can't see any sign of that in the code...

**Comment by erikjohnston** [comment]
Woops, looks like its `network_id` https://github.com/matrix-org/synapse/pull/1676/files#diff-0dca1a8726bd4f8d341ea3373b1048ddR182

**Comment by erikjohnston** [comment]
Will probably change it to `network_id` in the AS layer, and `instance_id` in the client layer, if that sounds good to you @lukebarnard1 ?

**Comment by erikjohnston** [comment]
@matrixbot retest this please

**Pull request merged**